src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -721,43 +716,76 @@ void ProcessVirtualMethod (MethodDefinition method)
 			}
 		}
 
-		void ProcessOverride (OverrideInformation overrideInformation)
+		/// <summary>
+		/// Returns true if the Override in <paramref name="overrideInformation"/> should be marked because it is needed by the base method.
+		/// Does not take into account if the base method is in a preserved scope.
+		/// </summary>
+		/// <param name="isDirectBase">True if the Override is the immediate override method in the type hierarchy.
+		/// False if Override is further removed (e.g. Override is a method on a derived type of a derived type of Base.DeclaringType)</param>
+		// TODO: Take into account a base method in preserved scope
+		// TODO: Move interface method marking logic here
+		bool ShouldMarkOverrideForBase (OverrideInformation overrideInformation, bool isDirectBase)
 		{
-			var method = overrideInformation.Override;
-			var @base = overrideInformation.Base;
-			if (!Annotations.IsMarked (method.DeclaringType))
-				return;
+			if (overrideInformation.IsOverrideOfInterfaceMember) {
+				_interfaceOverrides.Add ((overrideInformation, ScopeStack.CurrentScope));
+				return false;
+			}
 
-			if (Annotations.IsProcessed (method))
-				return;
+			var isInstantiated = Annotations.IsInstantiated (overrideInformation.Override.DeclaringType);
 
-			if (Annotations.IsMarked (method))
-				return;
+			bool markForOverride = !Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, overrideInformation.Override);
```

Comment:
Should not this "abort" the whole logic here?

Reply:
That's a good point, we can early return instead of or-ing markForOverride each time.